### DISCUSSION TITLE
wall placement scenarios and shadow placement shortcomings (no wall shadows)

### DISCUSSION THREAD

**Discussion by thibaud-be** [google/model-viewer#3998] [General]
Considering AR provides 2 scenarios floor/wall that (scene-viewer) calculates the shadows accordingly.
having modeleviewer only supporting floor shadowing creates a discontinuity with scene-viewer.
whenever presenting wall placement objects we typically have to desactivate modelviewer shadows and rely on opacity materials to fake them. the problem then being that in AR view ((scene-viewer) the opacity blended materials are not ignored and shadows are rendered atop the opacity blended shadow material. 
on android using the https link will use default webxr ar mode while using the http link will use scene-viewer
[https://modelec.doublerainbow.fr/](https://modelec.doublerainbow.fr/)
[http://modelec.doublerainbow.fr/](http://modelec.doublerainbow.fr/)
_clicking the switch will actually switch the light :)_

_using ar-placement="wall"`scene-viewer draws an "extra" shadow._
![scene-viewer-wall](https://user-images.githubusercontent.com/5654382/207043113-82388fb6-f232-4d92-9a58-3931d178ca58.jpg)
_using ar-placement="floor"_
![scene-viewer-floor](https://user-images.githubusercontent.com/5654382/207043109-c3c9f3b1-b6a9-4edd-8c02-b8b54267d920.jpg)

I'm not sur how to best workaround this issue, I thought of watching the ar-status event and replace the shadow material but I couldn't come up with any material that would prevent the backplane to be rendered at all.
using:
` { "name": "wall-shadow", "pbrMetallicRoughness": { "baseColorFactor": [1,1,1,0] },"alphaMode": "MASK","alphaCutoff": 1 },`
does not prevent the object to have rendered shadows in scene-viewer.
Although this should probably be considered a scene-viewer bug.  I can't help thinking that having a shadow system in model-viewer that allows both scenario (floor and wall) would be the best solution.

**Comment by elalish** [comment]
> Although this should probably be considered a scene-viewer bug. I can't help thinking that having a shadow system in model-viewer that allows both scenario (floor and wall) would be the best solution.

I'm having a little trouble following. You say there is a problem with scene viewer and you want a floor-wall shadow system in model-viewer itself, but that is exactly WebXR mode. Is that working properly? It's our default mode for lots of reasons like this.

**Reply by thibaud-be** [reply]
I've disregarded using webxr for several reasons:
- wall placement is desperately useless I couldn't get the thing to stick on any wall in my tests.
- there's a serious bug somewhere (on android 13) when trying to close the ar sometimes the entire browser freezes
- there are no shadow rendered at all nor a webxr one nor the one from the model itself

**Reply by elalish** [reply]
Thank you, that's very helpful feedback! Agreed that wall placement isn't great; we're working on that. I believe the close bug has been fixed; please try the latest chrome and report if it still repros. WebXR mode should be rendering a shadow, but I can repro from your link that it's not. That's definitely a bug.

**Comment by elalish** [comment]
@AdrianAtGoogle It looks like SceneViewer is making a baked (transparent) shadow plane cast a solid wall shadow. I haven't seen that for floor shadows - any idea what's going on?
